### ISSUE TITLE
Announce HTTP endpoint to the indexer 

### ISSUE THREAD

**Issue by olizilla** [web3-storage/w3up#786]
from @willscott 
> Announce HTTP endpoint as an extended family member to the indexer 
>   there's a multicodec to advertise for supporting http transfer that we'll want to have as an 'extended family member' of the eipfs provider's advertisement stream.

Today, only dag roots + paths can be fetched via our http gateway, and some older dags are not in R2.
...meanwhile e-ipfs can provide any block of any dag that has been uploaded to dotStorage.

Options considered:
- 🙅 do the work to make all CIDs available via http now, then announce http as an alternative provider for existing blocks
- 🙅 make a separate announcement for the blocks we have available over http
- ✅ announce http as an alternative provider now, even though we only support root cids today

By publishing the update to the existing announcement, we get a quick win of allowing saturn to fetch content via http where possible. Where we fail to provide blocks we have via http, the e-ipfs announcement will still work. 

Shortly afterwards we will backfill the set of dags that exist in S3 but not R2 to close that gap, and make a call on if we intend to support block level access over http.

We will announce http endpoint to IPNI as `freeway.dag.house`:
- we don't want to announce the internal `freeway.dag.haus` domain.
- `freeway.dag.house` pairs nicely with `elastic.dag.house`, 
- a separate domain may give us some sense of how much traffic is driven from IPNI discovery. 
- we don't want to use w3s.link here, as that would proxy some requests to dweb.link which doesn't support car-scope etc yet

```[tasklist]
### Tasks
- [x] Create domain alias `freeway.dag.house`^
- [ ] https://github.com/elastic-ipfs/publisher-lambda/pull/27
```

**Comment by vasco-santos** [comment]
I would say the end goal should be to announce all content. However, it is currently blocked on copying all our content into R2 (or not, it will redirect to dweb.link)

Depending on the timelines, we might accept the redirection (given it goes from dweb.link and might be actually fetched from other nodes that are not dweb.link). As the time goes, we will make content available in R2 and not redirect more.

**Comment by alanshaw** [comment]
...some options:

## Just make it work

Say that freeway serves ALL the blocks that E-IPFS serves.

* Expose an API to E-IPFS index data.
* If freeway doesn't have a block (not in R2), fallback to fetching index data from E-IPFS.
* Export directly from S3 buckets.

Locks us into providing _any_ block in DAGs our users have uploaded (not just root+path).

Need to look into the associated costs. We don't want this to come out more expensive than E-IPFS.

## Build a block index in buckets "CARbar"

To allow freeway to serve any block, we could build a big index in a bucket of all blocks in all CARs (like E-IPFS has in DynamoDB).

Index of `Block CID -> CAR CID(s)`. The CIDs of CAR file(s) tha contain the block.

```
bafy.../bagy...
```

Basically **DUDEWHERE** but for _all_ blocks not just the root.

## Start a new IPNI chain

This would be a chain for freeway, where we'd advertrise just the CIDs in DUDEWHERE i.e. the root CIDs.

It solves the problem of freeway not providing access to data via _any_ CID in the DAG (since we only advertise that we have the root), and it gives us the option to update our index later and add ALL the blocks.

It also solves the problem that not all data from S3 is also in R2.

**Comment by olizilla** [comment]
Copy everything to R2 + make a block to car index is my current favourite.

Would be neat if our block index also indicated the byte offset and length where it could be found, to save fetching a separate index of the car.

If we can build it so that we could deprecate the dynamo index and S3 copy of things, that'd be rad.

However, I think we want to land this asap, so in the meantime maybe we land "just make it work" first.

**Comment by vasco-santos** [comment]
As a note, a bucket to index at block level will likely be expensive and hit rate limits of R2. Think it might be better to use KV for this use case instead

**Comment by olizilla** [comment]
Current v0 proposal: 
- set up freeway.dag.house as domain alias (dag.haus is for internal things)
- announce freeway.dag.house as http provider to indexer

This gets us a quick win where content is fetched via root cid + path, which we assume is the significant majority of requests.

Once the announcements are updated, lassie and saturn will be able to attempt http retrieval for blocks we have. If fetching from us via http fails, the e-ipfs option is available. It's not ideal, but gets us something folks can use asap, and pushes saturn e-ipfs traffic to http where possible. 

Then we can decide on the longer piece of work on how best to index data for R2, and if we intend to support random block cid access over http.

The risk to be aware of with this is if we dont then do the extra work, we may start to get a reputation hit if say a block is available from us over http but then fail to serve it, but as far as I am aware such systems do not exist yet.

**Comment by willscott** [comment]
can confirm the reputation system does not quite exist yet :)

**Comment by alanshaw** [comment]
@vasco-santos do you mean for badbits? We need a gateway that doesn't redirect/race, and that supports `?car-scope`. Why don't we have firewall on freeway?

**Comment by olizilla** [comment]
Completed in https://github.com/elastic-ipfs/publisher-lambda/pull/27
